icu4c/source/i18n/tznames_impl.cpp

```diff
@@ -2275,6 +2276,10 @@ TZDBTimeZoneNames::getMetaZoneNames(const UnicodeString& mzID, UErrorCode& statu
         return nullptr;
     }
     mzIDKey[mzID.length()] = 0;
+    if (!uprv_isInvariantUString(mzIDKey, mzID.length())) {
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return nullptr;
+    }
```

Comment:
The ticket says there are "many places" in the API that assume the time zone ID is all ASCII but don't check for it, but this is the only substantive change in the PR.  Do all of the "many places" vector through here, or did you only fix one instance of the problem?

And is returning `U_ILLEGAL_ARGUMENT_ERROR` the right thing to do here?  The unit test appears to be calling a different API that doesn't return an error code.  What's going on between the functions the unit test is calling and this logic, and why is that the right thing to be doing?

My uninformed take is that you might be better off just converting the input time zone ID into an invariant-characters string and then using that value as the time zone ID.  If you get an invalid ID from doing that conversion, that's fine-- now it's just an invalid ID and you can handle that the way everything else handles an invalid ID.

Reply:
getMetaZoneNames is the internal function that many public surfaced API called into. That is why if we check there it fix many functions.

Reply:
the problem is in TZDBTimeZoneNames but that is an internal class. I add the unit test which eventually call into TZDBTimeZoneNames